### PR TITLE
Epson projector

### LINKED ISSUE #29: Epson Projector

### PR DESCRIPTION
**Pull request by dukeraphaelng**
This PR will:
- Resolve #29

### CODE REVIEW

File: drivers/epson/projector/esc_vp21.cr [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+require "placeos-driver/interface/powerable"
+require "placeos-driver/interface/switchable"
+
+module Epson; end
+
+module Epson::Projector; end
+
+# Documentation: documentation link
+
+class Epson::Projector::EscVp21 < PlaceOS::Driver
```
Comment (KimBurgess) [inline drivers/epson/projector/esc_vp21.cr:12]: I'm not familiar with the protocol for this device, but unless this is specific to an exact model, can probably make this class make a little more generic.

File: drivers/epson/projector/esc_vp21.cr [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+  include Interface::Muteable
+
+  enum Inputs
+    HDMI
+    HDBaseT
+  end
+
+  # Discovery Information
+  tcp_port 1024
+  descriptive_name "Epson Projectors"
```
Comment (KimBurgess) [inline drivers/epson/projector/esc_vp21.cr:23]:
```suggestion
  descriptive_name "Epson Projector"
```

File: drivers/epson/projector/esc_vp21.cr [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+  generic_name :Display
+
+  def on_load
+    transport.tokenizer = Tokenizer.new("\r\n")
+
+    self[:volume_min] = 0
+    self[:volume_max] = 255
+
+    self[:power] = 0
```
Comment (KimBurgess) [inline drivers/epson/projector/esc_vp21.cr:32]: Best to set any initial state based on a device query, rather than explicit value.

File: drivers/epson/projector/esc_vp21.cr [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+  end
+
+  def on_update
+  end
+
+  def connected
+    # Have to init comms
+    send("ESC/VP.net\x10\x03\x00\x00\x00\x00")
+    do_poll
+    schedule.every(52.seconds) { do_poll }
```
Comment (KimBurgess) [inline drivers/epson/projector/esc_vp21.cr:45]: In cases where you want to create a repeating schedule and have it run straight away too you can use `schedule.every(52.seconds, true) { ... }`

File: drivers/epson/projector/esc_vp21.cr [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+  def power?(**options, &block)
+    options[:emit] = block unless block.nil?
+    options[:name] = :power_state
+    do_send(:PWR, **options)
+  end
+
+  #
+  # Input selection
+  #
+  INPUTS = {
```
Comment (KimBurgess) [inline drivers/epson/projector/esc_vp21.cr:85]: You can encode these values within the existing Inputs enum.

File: drivers/epson/projector/esc_vp21.cr [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+  }
+  INPUT_LOOKUP = INPUTS.invert
+
+  def switch_to(input : Inputs)
+    do_send(:SOURCE, INPUT_LOOKUP[input], name: :inpt_source)
+    do_send(:SOURCE, name: :inpt_query)
+
+    logger.debug { "-- epson LCD, requested to switch to: #{input}" }
+    self[:input] = input # for a responsive UI
+    self[:MUTE] = false
```
Comment (KimBurgess) [inline drivers/epson/projector/esc_vp21.cr:97]:
```suggestion
    self[:mute] = false
```

File: drivers/epson/projector/esc_vp21.cr [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+  end
+
+  #
+  # Volume commands are sent using the inpt command
+  #
+  def volume(vol : Int32, **options)
+    vol = 0 if vol < 0
+    vol = 255 if vol > 255
```
Comment (KimBurgess) [inline drivers/epson/projector/esc_vp21.cr:105]: Take a look at [clamp](https://crystal-lang.org/api/0.35.1/Comparable.html#clamp(min,max)-instance-method) for this use case.

File: drivers/epson/projector/esc_vp21.cr [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+
+    # Seems to only return ":" for this command
+    self[:volume] = vol
+    self[:unmute_volume] = vol if vol > 0 # Store the "pre mute" volume, so it can be restored on unmute
```
Comment (KimBurgess) [inline drivers/epson/projector/esc_vp21.cr:109]: This can probably be stored as an internal variable rather than exposed state.

File: drivers/epson/projector/esc_vp21.cr [hunk 9]
```diff
@@ -0,0 +1,274 @@
...
+
+  # Mutes audio + video
+  def mute(
+    state : Bool = true,
+    index : Int32 | String = 0,
+    layer : MuteLayer = MuteLayer::AudioVideo
+  )
+    logger.debug { "-- epson Proj, requested mute state: #{state}" }
+
+    do_send(:MUTE, state, name: :video_mute) # Audio + Video
```
Comment (KimBurgess) [inline drivers/epson/projector/esc_vp21.cr:121]: You need to honour the `layer` param here.

File: drivers/epson/projector/esc_vp21.cr [hunk 10]
```diff
@@ -0,0 +1,274 @@
...
+
+  def unmute_audio(index : Int32 | String = 0)
+    mute_audio(false, index)
+  end
+
+  def input?
+    do_send(:SOURCE, name: :inpt_query, priority: 0)
+  end
+
+  ERRORS = {
```
Comment (KimBurgess) [inline drivers/epson/projector/esc_vp21.cr:145]: This looks like a good spot for an enum.

File: drivers/epson/projector/esc_vp21.cr [hunk 11]
```diff
@@ -0,0 +1,274 @@
...
+    19 => "13: power supply unit error",
+    20 => "14: exhaust shutter error",
+    21 => "15: obstacle detection error",
+    22 => "16: IF board discernment error",
+  }
+
+  #
+  # epson Response code
+  #
+  def received(data, resolve, command) # Data is default received as a string
```
Comment (KimBurgess) [inline drivers/epson/projector/esc_vp21.cr:172]: This method signature looks a bit off - the received callback is passed the data and a task object.

File: drivers/epson/projector/esc_vp21.cr [hunk 12]
```diff
@@ -0,0 +1,274 @@
...
+    logger.debug { "epson Proj sent: #{data}" }
+
+    if data == ":"
+      return :success
```
Comment (KimBurgess) [inline drivers/epson/projector/esc_vp21.cr:176]: Have a read here for now to mark a task result: https://github.com/PlaceOS/drivers/blob/master/docs/writing-a-driver.md#queue

File: drivers/epson/projector/esc_vp21.cr [hunk 13]
```diff
@@ -0,0 +1,274 @@
...
+    end
+
+    data = data.split(/=|\r:/)
+    case data[0].to_sym
```
Comment (KimBurgess) [inline drivers/epson/projector/esc_vp21.cr:180]: `to_sym` is a Ruby method. This will need some crystallising.

**Review by KimBurgess — CHANGES_REQUESTED** [review]
I've added a couple of initial comments - looks like there's still a little way to go but more than happy to keep some conversation happening in-line here as you have queries.

One thing I will recommend for both this driver and any others is that the focus should not be on porting them across verbatim. The old Ruby drivers may act as a good guide to highlight any device oddities, but not something to reproduce. Crystal is a language of its own and there's a lot of opportunity to build these neatly and in a way that uses the best of what crystals type system has to offer. Spend some time to grok how the old driver is implemented, then use this knowledge to help model the new driver in a way that makes sense for the tools at hand, yourself and others working on it in the future.

File: drivers/epson/projector/esc_vp21.cr (hunk 9 above)
Comment (dukeraphaelng) [reply drivers/epson/projector/esc_vp21.cr:121]: Sorry what do you mean by 'honour'?

File: drivers/epson/projector/esc_vp21.cr (hunk 3 above)
Comment (dukeraphaelng) [reply drivers/epson/projector/esc_vp21.cr:32]: Could you clarify what you mean by a 'device query'?

Comment (KimBurgess) [reply drivers/epson/projector/esc_vp21.cr:32]: Sending a command to the device that returns the device state. For example, it looks like you already have `power?` and `input?` implemented for this.

The reason for this is if a driver is stopped from backoffice, when it restarts you don't want to generate state change events _unless_ the state has actually changed. A nice pattern to follow is use the `connected` method to query any state that the driver is exposing. That way if the device drops offline or the driver stops when connectivity is re-established the driver always shows the correct device state.

File: drivers/epson/projector/esc_vp21.cr (hunk 9 above)
Comment (KimBurgess) [reply drivers/epson/projector/esc_vp21.cr:121]: At the moment if `mute(true, layer: MuteLayer::Audio)` is called, this method will activate the device video mute. You just need to check that param, then execute the appropriate action(s).

**Comment by dukeraphaelng** [comment]
> Having a quick read now. To provide some further reading on `**options`, that's a use of [splatting](https://crystal-lang.org/reference/syntax_and_semantics/splats_and_tuples.html).
> 
> TL;DR it can be used as a continent way to forward an arbitrary set of keyword arguments around. It was used a lot in the previous Ruby drivers, but would recommend avoiding it unless absolutely needed here as it can lead to some sloppy type signatures.

Hi @KimBurgess , upon going through this comment In Gab's PR I thought it'd be a good idea to implement `Hash` instead of `splat`. However, this is not allowed since `splatting` in drivers is mostly used in 

```
queue(**({
      name: command,
    }.merge(options))) do
...
end
```

and queue's method is defined in `placeos-driver/src/placeos-driver.cr` as follows (line 114-117)
```
  # Queuing
  def queue(**opts, &block : Task -> Nil)
    @__queue__.add(**opts, &block)
  end
```
which disallows anything but splats and NamedTuple in the argument.

**Comment by KimBurgess** [comment]
That `options` forwarding can stay out of these initial driver implementation, then can add this down the track when / where appropriate.

For some further info, you can't splat a hash in Crystal. You could in Ruby, but this breaks when we introduce types and the safety they provide. Within crystal you can think of a double splat as some syntactic sugar.

Say we have the following method:
```crystal
def foo(a : String, b : Int32)
  # ...
end
```
We can call this via
```crystal
foo a: "hello", b: 42
```
But if we also had those params ready to go somewhere else in a NameTuple, the following is equivalent.
```crystal
x = {a: "hello", b: 42}
foo **x
```

The also works in the opposite direction
```crystal
def bar(**kwargs)
  p! kwargs
end

bar a: "hello", b: 42
#=> {a: "hello", b: 42}
```

This is the same for `*var`, which expands or captures `Tuples` to/from positional params too.

File: drivers/epson/projector/esc_vp21.cr (hunk 9 above)
Comment (pkheav) [reply drivers/epson/projector/esc_vp21.cr]: @dukeraphaelng Something like this https://github.com/PlaceOS/drivers/blob/master/drivers/samsung/displays/mdc_protocol.cr#L124-L132

or for the epson projector example
https://github.com/PlaceOS/drivers/pull/62/commits/60e64c38c5e7d7e967e94d098db0356ba3c55dd7#diff-a9b9e76ba00d89e0a1d6dcd683831bc983f1e69e2ac6736f710d2f4909bb3ca3R110-R116

File: drivers/epson/projector/esc_vp21.cr (hunk 1 above)
Comment (pkheav) [reply drivers/epson/projector/esc_vp21.cr:5]: I think ESC/VP21 is just the name of the protocol or maybe group of projectors 
image
